packages/kbn-alerts-ui-shared/src/common/hooks/use_search_alerts_query.ts

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { useQuery } from '@tanstack/react-query';
+import type { ValidFeatureId } from '@kbn/rule-data-utils';
+import type {
+  MappingRuntimeFields,
+  QueryDslFieldAndFormat,
+  QueryDslQueryContainer,
+  SortCombinations,
+} from '@elastic/elasticsearch/lib/api/typesWithBodyKey';
+import type { DataPublicPluginStart } from '@kbn/data-plugin/public';
+import { catchError, filter, lastValueFrom, map, of } from 'rxjs';
+import { set } from '@kbn/safer-lodash-set';
+import type { UseQueryOptions } from '@tanstack/react-query/src/types';
+import type { Alert } from '@kbn/alerting-types';
+import type { RuleRegistrySearchRequest, RuleRegistrySearchResponse } from '@kbn/alerting-types';
+import { DEFAULT_ALERTS_PAGE_SIZE } from '../constants';
+import type { EsQuerySnapshot, LegacyField } from '../types';
+import { AlertsQueryContext } from '../contexts/alerts_query_context';
+
+export interface UseSearchAlertsQueryParams {
+  // Dependencies
+  /**
+   * Kibana data plugin, used to perform the query
+   */
+  data: DataPublicPluginStart;
+  // Parameters
+  /**
+   * Array of feature ids used for authorization and area-based filtering
+   */
+  featureIds: ValidFeatureId[];
+  /**
+   * ES query to perform on the affected alert indices
+   */
+  query?: Pick<QueryDslQueryContainer, 'bool' | 'ids'>;
+  /**
+   * The alert document fields to include in the response
+   */
+  fields?: QueryDslFieldAndFormat[];
+  /**
+   * Sort combinations to apply to the query
+   */
+  sort?: SortCombinations[];
+  /**
+   * Runtime mappings to apply to the query
+   */
+  runtimeMappings?: MappingRuntimeFields;
+  /**
+   * The page index to fetch
+   */
+  pageIndex?: number;
+  /**
+   * The page size to fetch
+   */
+  pageSize?: number;
+}
+
+export interface AlertsQueryData {
+  alerts: Alert[];
+  oldAlertsData: LegacyField[][];
+  ecsAlertsData: unknown[];
+  total: number;
+  querySnapshot?: EsQuerySnapshot;
+}
+
+export const searchAlertsQueryPrefix = ['alerts', 'searchAlerts'] as const;
+
+/**
+ * Query alerts
+ */
+export const useSearchAlertsQuery = (
+  { data, ...params }: UseSearchAlertsQueryParams,
+  options?: UseQueryOptions
+) => {
+  const {
+    featureIds,
+    fields,
+    query = {
+      bool: {},
+    },
+    sort = [
+      {
+        '@timestamp': 'desc',
+      },
+    ],
+    runtimeMappings,
+    pageIndex = 0,
+    pageSize = DEFAULT_ALERTS_PAGE_SIZE,
+  } = params;
+  return useQuery(
+    [...searchAlertsQueryPrefix, JSON.stringify(params)],
+    ({ signal }): Promise<AlertsQueryData> =>
+      lastValueFrom(
+        data.search
```

Comment:
`lastValueFrom` returns the last emitted value when the observable completes and, according to the [docs](https://rxjs.dev/guide/glossary-and-semantics#complete), when an observable completes:

> it will finalize [...] cleaning up resources

I usually keep [this](https://stackoverflow.com/a/41177163/6193024) as a reference when in doubt: in this case the observable is finite (otherwise the promise would never resolve) so I'm pretty sure it self-unsubscribes.

Reply:
Thank you @umbopepato!